spec/Section 3 -- Type System.md

```diff
@@ -962,21 +1442,113 @@ query withNullableVariable($var: String) {
 Note: The Validation section defines providing a nullable variable type to
 a non-null input type as invalid.
 
-**Non-Null type validation**
+**Type Validation**
 
 1. A Non-Null type must not wrap another Non-Null type.
 
 
 ## Directives
 
-A GraphQL schema includes a list of the directives the execution
-engine supports.
+DirectiveDefinition : Description? directive @ Name ArgumentsDefinition? on DirectiveLocations
+
+DirectiveLocations :
+  - `|`? DirectiveLocation
+  - DirectiveLocations | DirectiveLocation
+
+DirectiveLocation :
+  - ExecutableDirectiveLocation
+  - TypeSystemDirectiveLocation
+
+ExecutableDirectiveLocation : one of
+  `QUERY`
+  `MUTATION`
+  `SUBSCRIPTION`
+  `FIELD`
+  `FRAGMENT_DEFINITION`
+  `FRAGMENT_SPREAD`
+  `INLINE_FRAGMENT`
+
+TypeSystemDirectiveLocation : one of
+  `SCHEMA`
+  `SCALAR`
+  `OBJECT`
+  `FIELD_DEFINITION`
+  `ARGUMENT_DEFINITION`
+  `INTERFACE`
+  `UNION`
+  `ENUM`
+  `ENUM_VALUE`
+  `INPUT_OBJECT`
+  `INPUT_FIELD_DEFINITION`
+
+A GraphQL schema describes directives which are used to annotate various parts
+of a GraphQL document as an indicator that they should be evaluated differently
+by a validator, executor, or client tool such as a code generator.
 
 GraphQL implementations should provide the `@skip` and `@include` directives.
 
+GraphQL implementations that support the type system definition language must
+provide the `@deprecated` directive if representing deprecated portions of
+the schema.
+
+Directives must only be used in the locations they are declared to belong in.
+In this example, a directive is defined which can be used to annotate a
+fragment definition:
+
+```graphql example
+directive @example on FIELD
+
+fragment SomeFragment on SomeType {
+  field @example
+}
+```
+
+Directive locations may be defined with an optional leading `|` character to aid
+formatting when representing a longer list of possible locations:
+
+```graphql example
+directive @example on
+  | FIELD
+  | FRAGMENT_SPREAD
+  | INLINE_FRAGMENT
+```
+
+Directives can also be used to annotate the type system definition language
+as well, which can be a useful tool for supplying additional metadata in order
+to generate GraphQL execution services, produce client generated runtime code,
```

Comment:
> [...] produce client generated runtime code

But if I understand https://github.com/graphql/graphql-js/pull/746#issuecomment-301554231, directives won't generally be visible in client code. Or am I misreading this?

Reply:
Perhaps I'm just struggling with confusion over what clients can be expected to have access to when consuming a GraphQL service. Here's my understanding: a GraphQL API can either be defined using SDL or directly in code. So, it would seem to me that if directives would be visible to and usable by client code after consuming the SDL definition of the server, the same data would be accessible via introspection of a non-SDL schema. Or that a non-SDL schema could render to a generated SDL file (e.g. Sangria's [Schema materilialization](http://sangria-graphql.org/learn/#schema-materialization)) with the full expressiveness of an "a priori SDL" schema.

After reading @leebyron's comment, I started thinking, "Oh! Directives are used in SDL to guide the _implementation_ of a GraphQL schema, not to provide client-consumable metadata." Which is why that clause is throwing me off.